File: gameoflife/java/src/main/java/gameoflife/Population.java
package gameoflife;

import java.util.Collections;
import java.util.HashSet;
import java.util.Set;

import static java.util.stream.Collectors.toSet;


final class Population {

    private final Set<Cell> cells;

    Population(Set<Cell> cells) {

        this.cells = new HashSet<>(cells);
    }

    Set<Cell> survivors(Rule rule) {

        return cells.stream().filter(cell -> rule.test(neighborsInPopulation(cell))).collect(toSet());
    }


    Set<Cell> revitalized(Rule rule) {

        return cells.stream()
            .flatMap(cell -> cell.neighbors().stream().filter(neighbor -> !cells.contains(neighbor)))
            .filter(deadNeighbor -> rule.test(neighborsInPopulation(deadNeighbor)))
            .collect(toSet());
    }


    private int neighborsInPopulation(Cell cell) {

        return (int) cell.neighbors().stream().filter(cells::contains).count();
    }


    Set<Cell> cells() {

        return Collections.unmodifiableSet(cells);
    }
}

File: gameoflife/kotlin/src/main/kotlin/gameoflife/Population.kt
package gameoflife

class Population(cells: Set<Cell>) {

    companion object {
        val CONWAY_SURVIVAL_RULE = { n: Int -> n in 2..3 }
        val CONWAY_REVITALIZATION_RULE = { n: Int -> n == 3 }
    }

    val cells = cells.toSet()


    fun survivorCells() = cells.filter { CONWAY_SURVIVAL_RULE(neighborsInPopulation(it)) }.toSet()


    fun revitalizedCells() = cells
            .flatMap { cellInPopulation -> cellInPopulation.neighbors.filter { it !in cells } }
            .filter { deadNeighbor -> CONWAY_REVITALIZATION_RULE(neighborsInPopulation(deadNeighbor)) }
            .toSet()


    private fun neighborsInPopulation(cell: Cell) = cell.neighbors.filter { it in cells }.count()
}